### ISSUE TITLE
Segmentation Fault with Ocaml API

### ISSUE THREAD

**Issue by x3ent3nte** [Z3Prover/z3#878]
Hello

I am trying to make an Enumeration sort using the Ocaml api

open Z3

let _ = 
	let config = [("model", "true"); ("proof", "false")] in
	let context = Z3.mk_context config in
	let operation = Enumeration.mk_sort_s context "Operation" ["ADD"; "SUB"] in
	exit 0
		
Unfortunately, when I run it I get 
"Segmentation fault: 11"

**Comment by wintersteiger** [comment]
I can't reproduce that, it works  just fine for me. What kind of system are you seeing this on, what is your compilation command line, your precise version of ocaml, ocamlfind, etc? Does other Ocaml code using Z3 work or you?

**Comment by x3ent3nte** [comment]
I am using: 
OS X Sierra
Ocaml version 4.03.0
ocamlfind version 1.6.2
Z3 version 4.5.1 - 64 bit

My compilation command is:
ocamlfind ocamlopt -o ocaml_translator -package Z3 -package core -linkpkg -thread ocaml_translator.ml

I have successfully ran Z3's example ocaml programs.

The faulty program is a stripped down version of a larger program that worked until I used Enumeration.mk_sort 
This function seems to be causing problems for some reason.

**Comment by martin-neuhaeusser** [comment]
I could not reproduce the segmentation fault on Debian x64 (with OCaml 4.04.0+flambda) natively, and not using callgrind and cachegrind, either (I used commit #b70f1f of Z3's github repository).
Will try it on MacOS X in the next days.

**Comment by x3ent3nte** [comment]
I tried using Ubuntu 16.04 instead of OS X and I was able to call Enumeration.mk_sort once without it crashing. Unfortunately another Enumeration.mk_sort call caused a seg fault again. The call was
```
let condition = Enumeration.mk_sort context cond_sym [cond_eq; cond_ne; cond_cs; cond_cc; cond_mi; cond_pl; cond_vs; cond_vc; cond_hi; cond_ls; cond_ge; cond_lt; cond_gt; cond_le; cond_al]
```
When I shortened the list argument a seg fault was avoided. 
I tried increasing the stack size but that didn't fix the problem. 
The program I ran is below.

```
open Z3

let _ =

	let config = [("model", "true"); ("proof", "false")] in
	let context = Z3.mk_context config in

	let op_sym = Symbol.mk_string context "Operation" in
	let op_add = Symbol.mk_string context "ADD" in
	let op_sub = Symbol.mk_string context "SUB" in
	let op_mov = Symbol.mk_string context "MOV" in
	let op_mvn = Symbol.mk_string context "MVN" in
	let op_cmp = Symbol.mk_string context "CMP" in
	let op_and = Symbol.mk_string context "AND" in
	let cond_sym = Symbol.mk_string context "Condition" in
	let cond_eq = Symbol.mk_string context "EQ" in
	let cond_ne = Symbol.mk_string context "NE" in
	let cond_cs = Symbol.mk_string context "CS" in
	let cond_cc = Symbol.mk_string context "CC" in
	let cond_mi = Symbol.mk_string context "MI" in
	let cond_pl = Symbol.mk_string context "PL" in
	let cond_vs = Symbol.mk_string context "VS" in
	let cond_vc = Symbol.mk_string context "VC" in
	let cond_hi = Symbol.mk_string context "HI" in
	let cond_ls = Symbol.mk_string context "LS" in
	let cond_ge = Symbol.mk_string context "GE" in
	let cond_lt = Symbol.mk_string context "LT" in
	let cond_gt = Symbol.mk_string context "GT" in
	let cond_le = Symbol.mk_string context "LE" in
	let cond_al = Symbol.mk_string context "AL" in

	let operation = Enumeration.mk_sort context op_sym [op_add; op_sub; op_mov; op_mvn; op_cmp; op_and] in
	let condition = Enumeration.mk_sort context cond_sym [cond_eq; cond_ne; cond_cs; cond_cc; cond_mi; cond_pl; cond_vs; cond_vc; cond_hi; cond_ls; cond_ge; cond_lt; cond_gt; cond_le; cond_al] in 
	Printf.printf "Crashes before here\n";
	exit 0

```

It's a shame because I need to enumerate different symbols for cpu registers and condition mnemonics. Is there another way of doing this through the api?

**Comment by wintersteiger** [comment]
> Is there another way of doing this through the api?

You can use bit-vectors to encode the symbols. That avoids the use of the data-type theory that enums are based on. 

I'm trying to reproduce it on OSX with ocaml 4.03.0 at the moment, but don't have a verdict yet.

**Comment by wintersteiger** [comment]
Yep, both, native and bytcode, segfault on OSX 10.11.6 using OCaml 4.03.0.

**Comment by martin-neuhaeusser** [comment]
The same (with the extended example) on Debian x64.
I'll have a look into the OCaml bindings.

**Comment by wintersteiger** [comment]
The crash happens during Z3_inc_ref, so I suspect the reference counting is broken somewhere in the vicinity of mk_enumeration_sort.

**Comment by wintersteiger** [comment]
I think I figured it out. @martin-neuhaeusser : In z3native_stubs.c at line 1541:

```
  _a4_val = Val_emptylist;
  for (_i = _a2; _i > 0; _i--) {
    Z3_ast_plus _a4p = Z3_ast_plus_mk(ctx_p, (Z3_ast) _a4[_i]);
```

The last argument to `Z3_ast_plus_mk` should be `_a4[_i-1]` instead of `_a4[_i]` or the loop boundaries have to be shifted by one. It's likely that the same bug is present for all functions that have output arrays. Could you take a closer look at this?

**Comment by martin-neuhaeusser** [comment]
Yes, I am at the same position.
The array indices are shifted by one.

I’ll try to fix it now… just need to figure out where in the python script the code is generated – it’s a mess ;-)

Von: Christoph M. Wintersteiger [mailto:notifications@github.com]
Gesendet: Montag, 30. Januar 2017 16:54
An: Z3Prover/z3
Cc: Neuhaeusser, Martin (DF TI AT 2); Mention
Betreff: Re: [Z3Prover/z3] Segmentation Fault with Ocaml API (#878)


I think I figured it out. @martin-neuhaeusser<https://github.com/martin-neuhaeusser> : In z3native_stubs.c at line 1541:

  _a4_val = Val_emptylist;

  for (_i = _a2; _i > 0; _i--) {

    Z3_ast_plus _a4p = Z3_ast_plus_mk(ctx_p, (Z3_ast) _a4[_i]);

The last argument to Z3_ast_plus_mk should be _a4[_i-1] instead of _a4[_i] or the loop boundaries have to be shifted by one. It's likely that the same bug is present for all functions that have output arrays. Could you take a closer look at this?

**Comment by wintersteiger** [comment]
>  it’s a mess 

I know! Suggestions welcome :-)

**Comment by martin-neuhaeusser** [comment]
It is an off-by one error in accessing the output arrays in the OCaml stubs; I fixed the python code that generates the stub code.
With the change, @x3ent3nte 2nd example runs without a segmentation fault. I'll just run a subset of our internal benchmarks to check that I did not break anything and post a PR later.

**Comment by martin-neuhaeusser** [comment]
The (rather obvious) bug is fixed in #885 and things did not break in a quick test.

**Comment by wintersteiger** [comment]
Closing as fixed, thanks for the bug report!
